checker/src/main/java/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -492,9 +493,66 @@ public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror
         qualifier = qualifier.plusOffset(1);
         type.replaceAnnotation(convertUBQualifierToAnnotation(qualifier));
       }
+      if (imf.isStringIndexOf(tree)) {
+        // String#indexOf(String) and its variants that also take a String technically return
+        // (and are annotated as) @LTEqLengthOf the receiver. However, the result is always
+        // @LTLengthOf the receiver unless both the receiver and the target string are
+        // the empty string: "".indexOf("") returns 0, which isn't an index into "". So, this
+        // special case modifies the return type of these methods if either the parameter or
+        // the receiver is known (by the Value Checker) to not be the empty string. There are
+        // three ways the Value Checker might have that information: either string could have a
+        // @StringVal annotation whose value isn't "", either could have an @ArrayLen annotation
+        // whose value isn't zero, or either could have an @ArrayLenRange annotation whose from
+        // value is any positive integer.
+        ValueAnnotatedTypeFactory vatf =
+            ((UpperBoundAnnotatedTypeFactory) this.atypeFactory).getValueAnnotatedTypeFactory();
+        AnnotatedTypeMirror argType = vatf.getAnnotatedType(tree.getArguments().get(0));
+        AnnotatedTypeMirror receiverType = vatf.getReceiverType(tree);
+        if (definitelyIsNotTheEmptyString(argType, vatf)
+            || definitelyIsNotTheEmptyString(receiverType, vatf)) {
+          String receiverName = JavaExpression.getReceiver(tree).toString();
+          UBQualifier ltLengthOfReceiver = UBQualifier.createUBQualifier(receiverName, "0");
+          AnnotationMirror currentReturnAnno = type.getAnnotationInHierarchy(UNKNOWN);
+          UBQualifier currentUBQualifier =
+              UBQualifier.createUBQualifier(currentReturnAnno, (IndexChecker) checker);
+          UBQualifier result = currentUBQualifier.glb(ltLengthOfReceiver);
+          type.replaceAnnotation(convertUBQualifierToAnnotation(result));
+        }
+      }
       return super.visitMethodInvocation(tree, type);
     }
 
+    /**
+     * Returns true if the given Value Checker annotations guarantee that the annotated element is
+     * not the empty string.
+     *
+     * @param atm an annotated type from the Value Checker
+     * @param vatf the Value Annotated Type Factory
+     * @return true iff atm contains a {@code StringVal} annotation whose value isn't "", an {@code
+     *     ArrayLen} annotation whose value isn't 0, or an {@code ArrayLenRange} annotation whose
```

Comment:
Here, too, discuss containment (of "" or 0) rather than equality with that value.  The code gets this right.